components/script/dom/htmlscriptelement.rs

```diff
@@ -261,43 +388,76 @@ impl FetchResponseListener for ClassicContext {
+            (Err(err), _) | (_, Err(err)) => {
+                // Step 6, response is an error.
+                finish_fetching_a_classic_script(
+                    &*self.elem.root(),
+                    self.kind.clone(),
+                    self.url.clone(),
+                    Err(err.clone()),
+                );
+                return;
+            },
+            (Ok(_), Ok(_)) => {
+                let metadata = self.metadata.take().unwrap();
+
+                // Step 7.
+                let encoding = metadata
+                    .charset
+                    .and_then(|encoding| Encoding::for_label(encoding.as_bytes()))
+                    .unwrap_or(self.character_encoding);
+
+                // Step 8.
+                let (source_text, _, _) = encoding.decode(&self.data);
+                (source_text, metadata.final_url)
+            },
+        };
 
-            // Step 6.
-            let encoding = metadata
-                .charset
-                .and_then(|encoding| Encoding::for_label(encoding.as_bytes()))
-                .unwrap_or(self.character_encoding);
+        let elem = self.elem.root();
+        let global = elem.global();
+        let cx = global.get_cx();
+        let _ar = enter_realm(&*global);
+
+        let options =
+            unsafe { CompileOptionsWrapper::new(*cx, final_url.as_str().as_ptr() as *const i8, 1) };
+
+        let can_compile_off_thread =
+            unsafe { CanCompileOffThread(*cx, options.ptr as *const _, source_text.len()) };
+
+        if can_compile_off_thread {
+            let context = Box::into_raw(Box::new(OffThreadCompilationContext {
+                script_element: self.elem.clone(),
+                script_kind: self.kind.clone(),
+                final_url,
+                url: self.url.clone(),
+                task_source: global.dom_manipulation_task_source(),
+                canceller: global.task_canceller(TaskSourceName::DOMManipulation),
+                script_text: source_text.to_string(),
+            }));
 
-            // Step 7.
-            let (source_text, _, _) = encoding.decode(&self.data);
-            ScriptOrigin::external(
+            let source_text: Vec<u16> = source_text.encode_utf16().collect();
+
+            unsafe {
+                assert!(CompileOffThread(
+                    *cx,
+                    options.ptr as *const _,
+                    &mut transform_u16_to_source_text(&source_text) as *mut _,
+                    Some(off_thread_compilation_callback),
+                    context as *mut _ as *mut c_void,
+                ));
+            }
+        } else {
+            let load = ScriptOrigin::external(
                 Rc::new(DOMString::from(source_text)),
-                metadata.final_url,
-                self.fetch_options.clone(),
+                final_url.clone(),
+                ScriptFetchOptions::default_classic_script(&global),
```

Comment:
We should be using `self.fetch_options.clone()` here still.

Reply:
should we also store self.fetch_options in OffThreadCompilationContext?